### ISSUE TITLE
Websocket Reconnection Issues/Questions

### ISSUE THREAD

**Issue by usfngm** [tiagosiebler/binance#361]
Hey there. Thanks for the great API wrapper. I have a few questions that I didn't find the answer for.

TLDR:

- Does the reconnection mechanism keep trying forever? Or does it give up after the 3 times?
- Do I need to re-subscribe after getting a `reconnected` event?
   - `subscribeContinuousContractKlines` seems to be working fine 
   - `subscribeUsdFuturesUserDataStream` is the one I'm having my most problems with
- Also I keep getting the message below in my console, is this normal?

      (node:1485) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 close listeners added to [TLSSocket]. Use emitter.setMaxListeners() to increase limit

I have a strategy in mind that I'm coding and I'm indenting to leave this nodeJS app running 24/7. I know you have a reconnection mechanism in place for the `WebsocketClient` but is it guaranteed to try to reconnect forever if for some reason the computer lost its connection to the internet for hours?  Or does it give up after a few times? After reading other issues and peaking thru the code a little bit, I understood that (and correct me pls if I'm wrong) that it tries to reconnect 3 times but what happens after that? Does it give up and I have to reconnect it manually again? Because I had an internet outage for 2 hours and when the internet came back up, the WS did not reconnect. I just need to make sure this is the intended behavior.

Also regarding the Usd Futures User Data Stream. I understand it gets killed after one hour. Does it reconnect automatically after one hour? Or should I implement the reconnection logic myself? Because also, I was running my strategy today for a few hours and it took the wrong decisions because the user data stream wasn't streaming position updates even though I had no disconnects but when I checked the logs it kept saying.

     Completed keep alive cycle for listenKey(xxx) in market(usdm)

And do I need to re-subscribe to the user data stream upon receiving a `reconnected` event?

**Comment by tiagosiebler** [comment]
> * Does the reconnection mechanism keep trying forever? Or does it give up after the 3 times?

Yes, it keeps trying forever. The overall goal is for the ws client to worry about connectivity, so all you need to do is provide api credentials, ask for some topics and handle events as they arrive. The client will do the heavy lifting for you to ensure you stay connected.

> * Do I need to re-subscribe after getting a `reconnected` event?

No, that's why there are subscribe functions. The websocket client stores state on which topics you've subscribed to, so if a disconnect happens, it has all the info that it needs to replace a killed connection.

The `reconnected` event is useful for knowing you were disconnected for some time and may have missed some data. It's a good opportunity to check with the REST API, if you need to check for data that you missed since the last event (e.g. synchronising balance, orders, positions, etc).

>   * `subscribeContinuousContractKlines` seems to be working fine
>   * `subscribeUsdFuturesUserDataStream` is the one I'm having my most problems with

What kind of problem do you have with it, is it the event emitter leak below?

> * Also I keep getting the message below in my console, is this normal?
>   ```
>   (node:1485) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 close listeners added to [TLSSocket]. Use emitter.setMaxListeners() to increase limit
>   ```

This is not normal. It can happen if you have some recurring logic that keeps setting event listeners like:
`someEmitter.on('someEvent', (event) => .....);` 

Every time you call `someEmitter.on(...`, it creates a new event listener. I would check for any functions you have that are calling themselves, I would also check for any logic you have inside a setInterval callback. Make sure these event listeners are only created once and never again.

> I have a strategy in mind that I'm coding and I'm indenting to leave this nodeJS app running 24/7. I know you have a reconnection mechanism in place for the `WebsocketClient` but is it guaranteed to try to reconnect forever if for some reason the computer lost its connection to the internet for hours? Or does it give up after a few times? After reading other issues and peaking thru the code a little bit, I understood that (and correct me pls if I'm wrong) that it tries to reconnect 3 times but what happens after that? Does it give up and I have to reconnect it manually again? Because I had an internet outage for 2 hours and when the internet came back up, the WS did not reconnect. I just need to make sure this is the intended behavior.

I think you might be thinking of the keep-alive request. The user data stream is tied to a listen key. You have one listen key per account. Once it's created, by default it will expire after 1 hour, unless you call the keep alive function (even if you are still connected). The SDK will automatically keep a timer to send a keep-alive request on the listen key and keep track of it. If this request fails 3 times for whatever reason (e.g. disconnected wifi), it'll assume there's a problem with the listen key and/or network and a full reset is needed. 

A full reset means killing all timers (heartbeats + keep alive) for that user data stream, killing the idle connection if it's still there, asking the api for a new listen key, and reconnecting with the new listen key. This is handled automatically under the hood in this workflow:
https://github.com/tiagosiebler/binance/blob/master/src/websocket-client.ts#L950-L996

You shouldn't need to do anything to handle these lifecycle events, the networking is managed for you by the SDK. I originally had a lot of pain with some of the other open source binance SDKs (also other exchanges), so I spent a ton of time building out how this works to ensure I don't have to worry about dead connections that stopped receiving data. One easy way to test stability is by screwing with the connection on our machine (wifi off/on) and messing around with vpn connectivity (get some interesting networking behaviour while a vpn is connecting or stalled).

> Also regarding the Usd Futures User Data Stream. I understand it gets killed after one hour. Does it reconnect automatically after one hour? Or should I implement the reconnection logic myself? Because also, I was running my strategy today for a few hours and it took the wrong decisions because the user data stream wasn't streaming position updates even though I had no disconnects but when I checked the logs it kept saying.

This is actually handled for you, the user data stream typically sends an event if the listen key expires unexpectedly. I use that as a trigger to force a reconnect (kill old connection, ask for new listen key, reconnect with new listen key):
https://github.com/tiagosiebler/binance/blob/master/src/websocket-client.ts#L384-L392

I haven't seen this behaviour in a long time (in bots running for weeks 24/7). If you've found a way to reproduce the behaviour, I'd very much want to reproduce it too and investigate how to improve resilience even further.

Something I do want to ask is what kind of environment are you running on? Is it a local machine at home (e.g. connected to your wifi)? Is it a VPS? Dedicated server? cloud? Maybe that's a factor to reproducing this issue (most of my long-running testing is now on a dedicated server or vps).


> 
> ```
>  Completed keep alive cycle for listenKey(xxx) in market(usdm)
> ```
> 
> And do I need to re-subscribe to the user data stream upon receiving a `reconnected` event?

No, it's handled automatically. If a keep alive is successful, the same listen key remains active and nothing needs to change. If it fails or the response suggests a problem, the old connection should be killed and replaced under-the-hood. Since you're listening to events on the ws client, you don't need to do any actions to stay connected - the ws client will handle this "plumbing" for you automatically.

**Comment by usfngm** [comment]
Thank you for the detailed explanation. Yes I'm running my nodeJS app currently on a local machine using `pm2` on Node v20.5.1 with the plan of moving to a VPS once I did enough testing and confident enough that my code has no bugs. 

> This is not normal. It can happen if you have some recurring logic that keeps setting event listeners like:
someEmitter.on('someEvent', (event) => .....);

Yes I thought so but my code only adds these event listeners in a constructor and I only have one object with the constructor being called once. I built a wrapper on top of the WebSocketClient to make my custom events like `onCandleClose`, `onBalanceUpdate`, `onPositionUpdate`, etc...

Maybe I have to take a look once more or if I could provide you with a way to recreate this that would be helpful.

This is a very long shot 😅 and I'm not sure if this might be causing any issues but I'm using my Macbook M1 max in Clamshell mode with an external display and currently in the area I'm living in, we're experiencing power outages regularly. Whenever the power goes out, the macbook goes to sleep because in order for it to work with an external display and not sleep, it needs to be connected to a power source. I'm not sure how does the NodeJS process act when the computer sleeps and then wakes up again when the power is back. Maybe this is causing an issue with all the timers/heartbeat mechanism for reconnection. (idk?)

I looked at your code again and understood that if I call `subscribeUsdFuturesUserDataStream` it returns the existing connection (if it exists) and it does not create a new one. I tried setting a recurring call to `subscribeUsdFuturesUserDataStream` every 30 mins as a dirty way to ensure its alive. So far its working fine. (been running for 6 hours and so far it did not lose its connectivity to the user data stream.

I understand that I did not come to you with a "re-creatable" case here so I'd understand if you wanna close this issue. Thank you so much for your efforts and for this amazing API wrapper. I'll update this issue with my findings or what I did wrong once I figure it out (if the issue is still open) and thank you once more! 🙏

**Comment by tiagosiebler** [comment]
I think I've once reproduced a rare issue while my macbook was folded away / asleep & disconnected for an hour while commuting home. It's been a long time since I've seen it though, so I'm curious if there's still an issue lurking that only happens when a device is asleep & disconnected that long with the process still running. I'd say keep this open and see if we can pin down a clear pattern to reproduce. If there is still an issue it's a very rare one that doesn't happen in "normal" environments like a server, since I do also have processes on various servers running 24/7 uninterrupted with the user data stream...

**Comment by usfngm** [comment]
I think that was the issue. I deleted all the custom code I've written for "reconnectivity" and now only depend on the wrapper's internal mechanism. (I add the listeners only once, and call the subscribe methods only once)

To test this, I created a EC2 instance on AWS and ran the bot on it. Been up for 10 hours so far with no issues. That's far more uptime than I've ever had without issues on my local machine. 
<img width="1658" alt="image" src="https://github.com/tiagosiebler/binance/assets/13817039/12f80e5a-e9fa-4089-a787-50e6392cc4b0">

**Comment by tiagosiebler** [comment]
Good to hear! Do let me know if you pin down a way to reproduce that issue semi-consistently though, or if you notice any other weird behaviours!

**Comment by tiagosiebler** [comment]
Closing this for now. User data websocket handling improved significantly with #380 (v2.9.0), to also fix a rare issue with reconnecting after a listen key expired (after being disconnected for extended periods of time - e.g. with the laptop closed/asleep)). If you notice any issues or weird behaviour with the latest version, feel free to reopen or make a new issue to investigate.
